openmls/TEST-DESIGN.md

```diff
@@ -0,0 +1,175 @@
+       ```rust
+       let (mut alice_group, ..) = MlsGroup::try_init(&case, "Alice").unwrap();
+       let [(mut bob_group, ..), (mut charly_group, ..)] = alice_group.try_invite(&case, ["Bob", "Charly"]).unwrap();
+       ```
+    3. `MlsGroup::try_talk_to<const N: usize>(&mut self, case: &TestCase, others: &mut [MlsGroup ; N])` e.g.
+       ```rust
+       let (mut alice_group, ..) = MlsGroup::try_init(&case, "Alice").unwrap();
+       let [(mut bob_group, ..), (mut charly_group, ..)] = alice_group.try_invite(&case, ["Bob", "Charly"]).unwrap();
+       assert!(alice_group.try_talk_to(&case, &mut [bob_group, charly_group]).is_ok());
+       ```
+11. Fixtures: in order for being more flexible, they could just require a single `case` variable. It would look like
+    this:
+    ```rust
+    pub struct TestCase<T: OpenMlsCryptoProvider = OpenMlsRustCrypto> {
+        pub ciphersuite: Ciphersuite,
+        pub backend: T,
+        pub cfg: MlsGroupConfig,
+    }
+
+    #[apply(mls_test)]
+    fn should_bla_when_blabla(case: TestCase) {
+        let (mut alice_group, ..) = MlsGroup::try_init(&case, "Alice").unwrap();
+        let [(mut bob_group, ..), (mut charly_group, ..)] = alice_group.try_invite(&case, ["Bob", "Charly"]).unwrap();
+        assert!(alice_group.try_talk_to(&case, &mut [bob_group, charly_group]).is_ok());
+    }
+    ```
+    1. Does it make sense to keep the backend in fixtures ? I don't see openmls supporting anything else other
+       than `MemoryKeyStore`. Also should we keep evercrypt as an option or discontinue it ?
+    2. `MlsGroupConfig` is introduced to tune the `WireFormatPolicy` to make sure everything also works
+       in `PURE_CIPHERTEXT` mode.
+12. Test file layout
+    * Do not hesitate nesting your tests within mod. They provide better readability. They can be collapsed/expanded.
+      Rule of thumb could be `1 method = 1 mod` but it can be otherwise.
+    * Imports should live in the top `tests` mod to avoid duplicates. Nested mods should not have anything other
+      than `use super::*;`
+    * Within a nested mod you can avoid repeating the mod name e.g.
+      ~~`method_a_should_bla_when_blabla`~~ -> `should_bla_when_blabla`
+
+```rust
+impl MyStruct {
+    pub fn method_a() {}
+    // snip
+    fn method_z() {}
+}
+
+// No `#[cfg(features = "test-utils")]`
+
+#[test]
+mod tests {
+    use super::*;
+    use crate::prelude::*; // if required
+
+    mod a {
+        use super::*; // <- no other imports here. 
+        // If one is required, add it to tests mod. This avoids duplicates.
+
+        #[test]
+        fn should_bla_when_blabla_1() {}
+
+        // snip
+        #[test]
+        fn should_bla_when_blabla_n() {}
+    }
+
+    // snip
+
+    mod z {
+        use super::*;
+
+        #[test]
+        fn should_bla_when_blabla_1() {}
+
+        // snip
+        #[test]
+        fn should_bla_when_blabla_n() {}
+    }
+}
+```
+
+* Crate layout
+    * unit tests in `src` in the file they relate to
```

Comment:
As I said above, module should be preferred but both is fine to do.

Reply:
You mean unit tests in separate files ? Like currently ?